gcloud/datastore/query.py

```diff
@@ -140,441 +243,266 @@ def filter(self, property_name, operator, value):
+        ancestor_filter.property.name = '__key__'
+        ancestor_filter.operator = datastore_pb.PropertyFilter.HAS_ANCESTOR
+        ancestor_filter.value.key_value.CopyFrom(ancestor_pb)
 
-        for prop in properties:
-            property_order = clone._pb.order.add()
+    for property_name, operator, value in query.filters:
+        pb_op_enum = query.OPERATORS.get(operator)
 
-            if prop.startswith('-'):
-                property_order.property.name = prop[1:]
-                property_order.direction = property_order.DESCENDING
-            else:
-                property_order.property.name = prop
-                property_order.direction = property_order.ASCENDING
+        # Add the specific filter
+        property_filter = composite_filter.filter.add().property_filter
+        property_filter.property.name = property_name
+        property_filter.operator = pb_op_enum
 
-        return clone
+        # Set the value to filter on based on the type.
+        if property_name == '__key__':
+            key_pb = value.to_protobuf()
+            property_filter.value.key_value.CopyFrom(
+                helpers._prepare_key_for_request(key_pb))
+        else:
+            helpers._set_protobuf_value(property_filter.value, value)
 
-    def projection(self, projection=None):
-        """Adds a projection to the query.
+    if not composite_filter.filter:
+        pb.ClearField('filter')
 
-        This is a hybrid getter / setter, used as::
+    for prop in query.order:
+        property_order = pb.order.add()
 
-          >>> query = Query('Person')
-          >>> query.projection()  # Get the projection for this query.
-          []
-          >>> query = query.projection(['name'])
-          >>> query.projection()  # Get the projection for this query.
-          ['name']
+        if prop.startswith('-'):
+            property_order.property.name = prop[1:]
+            property_order.direction = property_order.DESCENDING
+        else:
+            property_order.property.name = prop
+            property_order.direction = property_order.ASCENDING
 
-        :type projection: sequence of strings
-        :param projection: Each value is a string giving the name of a
-                           property to be included in the projection query.
+    for group_by_name in query.group_by:
+        pb.group_by.add().name = group_by_name
 
-        :rtype: :class:`Query` or `list` of strings.
-        :returns: If no arguments, returns the current projection.
-                  If a projection is provided, returns a clone of the
-                  :class:`Query` with that projection set.
-        """
-        if projection is None:
-            return [prop_expr.property.name
-                    for prop_expr in self._pb.projection]
+    return pb
 
-        clone = self._clone()
 
-        # Reset projection values to empty.
-        clone._pb.ClearField('projection')
+class _Iterator(object):
```

Comment:
Hmm, we don't follow that pattern in ``storage``.  It's private because we don't expect users to construct one directly (rather than by calling 'fetch()', or iterating the query itself).